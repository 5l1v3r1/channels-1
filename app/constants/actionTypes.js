// @flow
const actionTypes = {
  FETCH_ACCESS_TOKEN_REQUEST: 'FETCH_ACCESS_TOKEN_REQUEST',
  FETCH_ACCESS_TOKEN_SUCCESS: 'FETCH_ACCESS_TOKEN_SUCCESS',
  FETCH_ACCESS_TOKEN_FAILURE: 'FETCH_ACCESS_TOKEN_FAILURE',
  FETCH_CHANNELS_REQUEST: 'FETCH_CHANNELS_REQUEST',
  FETCH_CHANNELS_SUCCESS: 'FETCH_CHANNELS_SUCCESS',
  FETCH_CHANNELS_FAILURE: 'FETCH_CHANNELS_FAILURE',
  UPDATE_ACTIVE_CHANNEL: 'UPDATE_ACTIVE_CHANNEL',
  FETCH_VIDEOS_REQUEST: 'FETCH_VIDEOS_REQUEST',
  FETCH_VIDEOS_SUCCESS: 'FETCH_VIDEOS_SUCCESS',
  FETCH_VIDEOS_FAILURE: 'FETCH_VIDEOS_FAILURE'
};

export default actionTypes;
