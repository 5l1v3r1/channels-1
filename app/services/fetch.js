// @flow
export const Methods = {
  GET: 'GET',
  POST: 'POST'
};

export function authRequest(uri: string): Promise<any> {
  const [url, body] = uri.split('?');
  const init = {
    method: Methods.POST,
    headers: { 'Content-Type': 'application/x-www-form-urlencoded' },
    body
  };
  return fetch(url, init)
    .then(res => {
      if (!res.ok) {
        return Promise.reject();
      }
      return res.json();
    });
}

export function getRequest(uri: string): Promise<any> {
  const init = {
    method: Methods.GET,
    headers: { 'Content-Type': 'application/json' }
  };
  return fetch(uri, init)
    .then(res => {
      if (!res.ok) {
        return Promise.reject();
      }
      return res.json();
    });
}

export function postRequest(uri: string, body?: {}): Promise<any> {
  const init = body ?
  {
    method: Methods.POST,
    headers: { 'Content-Type': 'application/json' },
    body: JSON.stringify(body)
  } :
  {
    method: Methods.POST,
    headers: { 'Content-Type': 'application/x-www-form-urlencoded' }
  };
  return fetch(uri, init)
    .then(res => {
      if (!res.ok) {
        return Promise.reject();
      }
      return res;
    });
}
