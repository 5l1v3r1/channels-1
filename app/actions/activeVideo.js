// @flow
import actionTypes from '../constants/actionTypes';
import type { ThunkAction, Dispatch, GetState } from '../constants/typeAliases';
import RatingType from '../dataTypes/ratingType';

export function updateActiveVideo(video: any): ThunkAction {
  return (dispatch: Dispatch, getState: GetState) => {
    const { id } = getState().activeVideo;
    if (id !== video.id) {
      dispatch({
        type: actionTypes.UPDATE_ACTIVE_VIDEO,
        payload: video
      });
    }
  };
}

export function updateActiveVideoCounts(prevState: RatingType, rating: string): ThunkAction {
  return (dispatch: Dispatch, getState: GetState) => {
    let likeCount = Number(getState().activeVideo.likeCount);
    let dislikeCount = Number(getState().activeVideo.dislikeCount);
    switch (rating) {
      case 'like':
        likeCount = prevState.like ? likeCount - 1 : likeCount + 1;
        dislikeCount = prevState.dislike ? dislikeCount - 1 : dislikeCount;
        break;
      case 'dislike':
        likeCount = prevState.like ? likeCount - 1 : likeCount;
        dislikeCount = prevState.dislike ? dislikeCount - 1 : dislikeCount + 1;
        break;
      default:
        likeCount = prevState.like ? likeCount - 1 : likeCount;
        dislikeCount = prevState.dislike ? dislikeCount - 1 : dislikeCount;
    }
    dispatch({
      type: actionTypes.UPDATE_ACTIVE_VIDEO,
      payload: {
        likeCount: String(likeCount),
        dislikeCount: String(dislikeCount)
      }
    });
  };
}
