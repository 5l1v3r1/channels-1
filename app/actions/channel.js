// @flow
import actionTypes from '../constants/actionTypes';
import type { Action } from '../constants/typeAliases';

export function selectChannel(id: string): Action {
  return {
    type: actionTypes.SELECT_CHANNEL,
    payload: id
  };
}
