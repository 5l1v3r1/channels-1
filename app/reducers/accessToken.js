// @flow
import { combineReducers } from 'redux';
import actionTypes from '../constants/actionTypes';
import type { Action } from '../constants/typeAliases';

function token(state: string = '', { type, payload }: Action): string {
  switch (type) {
    case actionTypes.FETCH_ACCESS_TOKEN_SUCCESS:
      return payload;
    default:
      return state;
  }
}

function isFetching(state: boolean = false, { type }: Action): boolean {
  switch (type) {
    case actionTypes.FETCH_ACCESS_TOKEN_REQUEST:
      return true;
    case actionTypes.FETCH_ACCESS_TOKEN_SUCCESS:
    case actionTypes.FETCH_ACCESS_TOKEN_FAILURE:
      return false;
    default:
      return state;
  }
}


function error(state: boolean = false, { type }: Action): boolean {
  switch (type) {
    case actionTypes.FETCH_ACCESS_TOKEN_REQUEST:
    case actionTypes.FETCH_ACCESS_TOKEN_SUCCESS:
      return false;
    case actionTypes.FETCH_ACCESS_TOKEN_FAILURE:
      return true;
    default:
      return state;
  }
}

const accessToken = combineReducers({
  token,
  isFetching,
  error
});

export default accessToken;
