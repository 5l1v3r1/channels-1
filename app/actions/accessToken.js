// @flow
import actionTypes from '../constants/actionTypes';
import type { Action, Dispatch } from '../constants/typeAliases';
import { authRequest } from '../services/fetch';
import { getAccessTokenUri } from '../services/uriGenerator';

type ThunkAction = (dispatch: Dispatch) => any;

export function fetchAccessToken(): ThunkAction {
  return (dispatch: Dispatch) => {
    dispatch(requestAccessToken());
    const uri = getAccessTokenUri();
    return authRequest(uri)
      .then(json => dispatch(receiveAccessToken(json.access_token)))
      .catch(() => dispatch(accessTokenError()));
  };
}

export function requestAccessToken(): Action {
  return {
    type: actionTypes.FETCH_ACCESS_TOKEN_REQUEST
  };
}

export function receiveAccessToken(token: string): Action {
  return {
    type: actionTypes.FETCH_ACCESS_TOKEN_SUCCESS,
    payload: token
  };
}

export function accessTokenError(): Action {
  return {
    type: actionTypes.FETCH_ACCESS_TOKEN_FAILURE
  };
}
