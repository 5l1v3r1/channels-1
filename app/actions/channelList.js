// @flow
import actionTypes from '../constants/actionTypes';
import type { Action, Dispatch, AccessTokenState } from '../constants/typeAliases';
import { apiRequest, Methods } from '../services/fetch';
import { getChannelsUri, getSubscriptionsUri } from '../services/uriGenerator';

type GetState = () => { accessToken: AccessTokenState };

type ThunkAction = (dispatch: Dispatch, getState: GetState) => any;

export function fetchChannels(): ThunkAction {
  return (dispatch: Dispatch, getState: GetState) => {
    dispatch(requestChannels());
    const { accessToken } = getState().accessToken;
    const subscriptions = fetchSubscriptions(accessToken);
    subscriptions.then(json => {
      if (json.items == null) {
        return dispatch(receiveChannels([]));
      }
      const subscriptions = json.items.map(item => item.snippet.resourceId.channelId).join();
      const uri = getChannelsUri(accessToken, subscriptions);
      return apiRequest(uri, Methods.GET)
        .then(json => {
          if (json.items == null) {
            return dispatch(receiveChannels([]));
          }
          dispatch(receiveChannels(json));
        });
    })
    .catch(() => dispatch(channelsError()));
  };
}

function fetchSubscriptions(accessToken: string): Promise<any> {
  const uri = getSubscriptionsUri(accessToken);
  return apiRequest(uri, Methods.GET);
}

function requestChannels(): Action {
  return {
    type: actionTypes.REQUEST_CHANNELS
  };
}

function receiveChannels(json: any): Action {
  return {
    type: actionTypes.RECEIVE_CHANNELS,
    payload: json.items
  };
}

function channelsError(): Action {
  return {
    type: actionTypes.CHANNELS_ERROR
  };
}
