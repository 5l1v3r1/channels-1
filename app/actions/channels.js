// @flow
import actionTypes from '../constants/actionTypes';
import type { Action, ThunkAction, Dispatch, GetState } from '../constants/typeAliases';
import { apiRequest, Methods } from '../services/fetch';
import { getChannelsUri, getSubscriptionsUri } from '../services/uriGenerator';

export function fetchChannels(): ThunkAction {
  return (dispatch: Dispatch, getState: GetState) => {
    dispatch(requestChannels());
    const { token } = getState().accessToken;
    const subscriptions = fetchSubscriptions(token);
    return subscriptions.then(json => {
      if (json.items == null) {
        return dispatch(receiveChannels([]));
      }
      const subscriptions = json.items.map(item => item.snippet.resourceId.channelId).join();
      const uri = getChannelsUri(token, subscriptions);
      return apiRequest(uri, Methods.GET)
        .then(json => {
          if (json.items == null) {
            return dispatch(receiveChannels([]));
          }
          return dispatch(receiveChannels(json.items));
        });
    })
    .catch(() => dispatch(channelsError()));
  };
}

export function fetchSubscriptions(token: string): Promise<any> {
  const uri = getSubscriptionsUri(token);
  return apiRequest(uri, Methods.GET);
}

export function requestChannels(): Action {
  return {
    type: actionTypes.FETCH_CHANNELS_REQUEST
  };
}

export function receiveChannels(items: any): Action {
  return {
    type: actionTypes.FETCH_CHANNELS_SUCCESS,
    payload: items
  };
}

export function channelsError(): Action {
  return {
    type: actionTypes.FETCH_CHANNELS_FAILURE
  };
}

export function updateActiveChannel(channelId: string): ThunkAction {
  return (dispatch: Dispatch, getState: GetState) => {
    const { activeId } = getState().channels;
    if (activeId !== channelId) {
      dispatch({
        type: actionTypes.UPDATE_ACTIVE_CHANNEL,
        payload: channelId
      });
    }
  };
}
