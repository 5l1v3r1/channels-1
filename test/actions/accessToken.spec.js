/* eslint promise/always-return: 0 */
import configureMockStore from 'redux-mock-store';
import thunk from 'redux-thunk';
import * as accessTokenActions from '../../app/actions/accessToken';
import actionTypes from '../../app/constants/actionTypes';
import * as fetch from '../../app/services/fetch';

const middlewares = [thunk];
const mockStore = configureMockStore(middlewares);

jest.mock('../../app/services/uriGenerator');

describe('accessToken actions', () => {

  const token = 'ya29.GlyrBCJQJoIYFzocIunVN-CfjQZMG4oyVuAB6v_x_Z3FRnViyPy_deqRdwSAipQtKc9Nb2RudM9UISwI8SGNXxsJ1t3QHddeCdnoCjsM_vhLa9FlFVqMN_seI7oljg';

  it('creates FETCH_ACCESS_TOKEN_SUCCESS when fetching access token has been resolved', () => {
    fetch.authRequest = jest.fn(() => Promise.resolve({ access_token: token }));

    const expectedActions = [
      { type: actionTypes.FETCH_ACCESS_TOKEN_REQUEST },
      { type: actionTypes.FETCH_ACCESS_TOKEN_SUCCESS, payload: token }
    ];
    const store = mockStore({});

    return store.dispatch(accessTokenActions.fetchAccessToken()).then(() => {
      expect(store.getActions()).toEqual(expectedActions);
    });
  });

  it('creates FETCH_ACCESS_TOKEN_FAILURE when fetching access token has been rejected', () => {
    fetch.authRequest = jest.fn(() => Promise.reject());

    const expectedActions = [
      { type: actionTypes.FETCH_ACCESS_TOKEN_REQUEST },
      { type: actionTypes.FETCH_ACCESS_TOKEN_FAILURE }
    ];
    const store = mockStore({});

    return store.dispatch(accessTokenActions.fetchAccessToken()).then(() => {
      expect(store.getActions()).toEqual(expectedActions);
    });
  });

  it('should create an action to request an access token', () => {
    const expectedAction = {
      type: actionTypes.FETCH_ACCESS_TOKEN_REQUEST
    };

    expect(accessTokenActions.requestAccessToken()).toEqual(expectedAction);
  });

  it('should create an action to receive an access token', () => {
    const expectedAction = {
      type: actionTypes.FETCH_ACCESS_TOKEN_SUCCESS,
      payload: token
    };

    expect(accessTokenActions.receiveAccessToken(token)).toEqual(expectedAction);
  });

  it('should create an action to handle an error', () => {
    const expectedAction = {
      type: actionTypes.FETCH_ACCESS_TOKEN_FAILURE
    };

    expect(accessTokenActions.accessTokenError()).toEqual(expectedAction);
  });
});
