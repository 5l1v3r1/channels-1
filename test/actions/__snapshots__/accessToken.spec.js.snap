// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`accessToken actions creates FETCH_ACCESS_TOKEN_FAILURE when fetching access token has been rejected 1`] = `
Array [
  Object {
    "type": "FETCH_ACCESS_TOKEN_REQUEST",
  },
  Object {
    "type": "FETCH_ACCESS_TOKEN_FAILURE",
  },
]
`;

exports[`accessToken actions creates FETCH_ACCESS_TOKEN_SUCCESS when fetching access token has been resolved 1`] = `
Array [
  Object {
    "type": "FETCH_ACCESS_TOKEN_REQUEST",
  },
  Object {
    "payload": "ya29.GlyrBCJQJoIYFzocIunVN-CfjQZMG4oyVuAB6v_x_Z3FRnViyPy_deqRdwSAipQtKc9Nb2RudM9UISwI8SGNXxsJ1t3QHddeCdnoCjsM_vhLa9FlFVqMN_seI7oljg",
    "type": "FETCH_ACCESS_TOKEN_SUCCESS",
  },
]
`;

exports[`accessToken actions should create an action to handle an error 1`] = `
Object {
  "type": "FETCH_ACCESS_TOKEN_FAILURE",
}
`;

exports[`accessToken actions should create an action to receive an access token 1`] = `
Object {
  "payload": "ya29.GlyrBCJQJoIYFzocIunVN-CfjQZMG4oyVuAB6v_x_Z3FRnViyPy_deqRdwSAipQtKc9Nb2RudM9UISwI8SGNXxsJ1t3QHddeCdnoCjsM_vhLa9FlFVqMN_seI7oljg",
  "type": "FETCH_ACCESS_TOKEN_SUCCESS",
}
`;

exports[`accessToken actions should create an action to request an access token 1`] = `
Object {
  "type": "FETCH_ACCESS_TOKEN_REQUEST",
}
`;
